### ISSUE TITLE
Giza: QN processor crash on delete distribution bucket

### ISSUE THREAD

**Issue by traumschule** [Joystream/joystream#3457]
Missing runtime/processor checks when deleting distribution buckets.

Deleted distribution buckets and changes weren't reflected on https://joystreamstats.live/distribution
```
$ /home/joystream/joystream/node_modules/.bin/joystream-distributor leader:delete-bucket -B 5:1 -y
2022-03-17 18:18:50:1850 CLI info: Deleting distribution bucket...
{ 
"bucketId": {
"distribution_bucket_family_id": "5",
"distribution_bucket_index": "1" 
}
}
2022-03-17 18:18:50:1850 SubstrateApi info: Sending storage.deleteDistributionBucket extrinsic from 5EEucXxsz1jZhF5HwQ38cNn5LuZa7RquHquzzyKfnVc1trMy
2022-03-17 18:18:54:1854 CLI info: Bucket succesfully deleted! 
Done in 8.55s. 
```
processor crashed:
```
INFO: Last block: 4911667       : 23629 blocks behind
INFO: Starting the event queue
INFO: Processing block: 0004911669-aca04, events count: 1 
ERROR: Stopping the proccessor due to errors: {
  "message": "update or delete on table \"distribution_bucket\" violates foreign key constraint \"FK_678dc5427cdde0cd4fef2c07a43\" on table \"distribution_bucket_operator\"",
  "length": 370,
  "name": "QueryFailedError",
  "severity": "ERROR",
  "code": "23503",
  "detail": "Key (id)=(5:1) is still referenced from table \"distribution_bucket_operator\".",
  "schema": "public",
  "table": "distribution_bucket_operator",
  "constraint": "FK_678dc5427cdde0cd4fef2c07a43",
  "file": "ri_triggers.c",
  "line": "2476",
  "routine": "ri_ReportViolation",
  "query": "DELETE FROM \"distribution_bucket\" WHERE \"id\" = $1",
  "parameters": [
    "5:1"
  ]
} name: QueryFailedError, message: update or delete on table "distribution_bucket" violates foreign key constraint "FK_678dc5427cdde0cd4fef2c07a43" on table "distribution_bucket_operator", stack: QueryFailedError: update or delete on table "distribution_bucket" violates foreign key constraint "FK_678dc5427cdde0cd4fef2c07a43" on table "distribution_bucket_operator"
    at new QueryFailedError (/joystream/node_modules/typeorm/error/QueryFailedError.js:12:28)
    at PostgresQueryRunner.<anonymous> (/joystream/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:248:31)
    at step (/joystream/node_modules/tslib/tslib.js:143:27)
    at Object.throw (/joystream/node_modules/tslib/tslib.js:124:57)
    at rejected (/joystream/node_modules/tslib/tslib.js:115:69)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
ERROR: {} name: Error, message: QueryFailedError: update or delete on table "distribution_bucket" violates foreign key constraint "FK_678dc5427cdde0cd4fef2c07a43" on table "distribution_bucket_operator", stack: Error: QueryFailedError: update or delete on table "distribution_bucket" violates foreign key constraint "FK_678dc5427cdde0cd4fef2c07a43" on table "distribution_bucket_operator"
    at MappingsProcessor.processingLoop (/joystream/node_modules/@joystream/hydra-processor/lib/process/MappingsProcessor.js:75:23)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async Promise.all (index 1)
INFO: Shutting down...
    at async MappingsProcessor.start (/joystream/node_modules/@joystream/hydra-processor/lib/process/MappingsProcessor.js:25:9)
    at async ProcessorRunner.process (/joystream/node_modules/@joystream/hydra-processor/lib/start/ProcessorRunner.js:47:9)
    at async Run.run (/joystream/node_modules/@joystream/hydra-processor/lib/commands/run.js:27:13)
INFO: Closing the database connection...
    at async Run._run (/joystream/node_modules/@oclif/command/lib/command.js:43:20)
    at async Config.runCommand (/joystream/node_modules/@oclif/config/lib/config.js:173:24)
    at async Main.run (/joystream/node_modules/@oclif/command/lib/main.js:27:9)
    at async Main._run (/joystream/node_modules/@oclif/command/lib/command.js:43:20)
error Command failed with exit code 1.
```
Noticed because contributors [failed to upload videos](https://discord.com/channels/811216481340751934/859752198166282280/954375219848568913).

QN:
- https://hydra.joystream.org/graphql
- https://ipfs.joystreamstats.live/graphql

**Comment by mnaamani** [comment]
Investigating..

The runtime only successfully removes the distribution bucket as long as  no bags and no distributors are assigned to the bucket. So the query-node is not explicitly removing distributors on handling the `DistributionBucketDeleted` event. But clearly from error above there are still distributors referencing the bucket this the query-node is attempting to remove. Will check the implementation of the handler for `DistributionBucketOperatorRemoved`.

**Comment by mnaamani** [comment]
The distributiron operator that is still in the database is:

```
{
        "workerId": 14,
        "distributionBucketId": "5:1",
        "status": "INVITED",
        "distributionBucket": {
          "id": "5:1",
          "distributing": false
        }
      },
```

The handler for `DistributionBucketOperatorRemoved` does try to delete the operator. It is not clear if there was ever an event for this operator. Is the fact that the status of this operator is still set to `Invited` significant.. ? Will continue to investigate.

**Comment by traumschule** [comment]
Hint: the worker was [fired in block #4908750](https://pioneer.joystreamstats.live/#/explorer/query/4908750).

**Comment by mnaamani** [comment]
What I can see is that the operator also did not accept the invitation. Can you confirm that?

The runtime allows the bucket to be deleted, it checks only if there are "active" operators, ignoring any invited operators. The query-node mappings do not take this into account. So The fix as far as I can tell is the handler for `DistributionBucketDeleted` should look at the 'pernding_invitations' operators and delete them.

**Comment by mnaamani** [comment]
> Hint: the worker was [fired in block #4908750](https://pioneer.joystreamstats.live/#/explorer/query/4908750).

Interesting, as far as I can see from the handler for workers being terminated, it does not update the "operator" state. I have to investigate if there is any integration between working group and storage pallets at the runtime level when it comes to handling worker terminations. But I think the fix in previous comment is the correct fix.

**Comment by traumschule** [comment]
> What I can see is that the operator also did not accept the invitation. Can you confirm that?

That seems to be right, could not find `DistributionBucketInvitationAccepted` for worker 14 in my local events cache.
Inviation: https://pioneer.joystreamstats.live/#/explorer/query/4510284

**Comment by mnaamani** [comment]
I prepared a fix https://github.com/Joystream/joystream/pull/3458 can you test on your node @traumschule

**Comment by mnaamani** [comment]
I suggest you cherry-pick the fix only or you will need to restart processor to process from starting block.

**Comment by bedeho** [comment]
Can we have a TLDR on this?

**Comment by mnaamani** [comment]
I just deployed to live network.

**Comment by mnaamani** [comment]
> Can we have a TLDR on this?

A distribution bucket struct:
```rust
pub struct DistributionBucketRecord<WorkerId: Ord> {
    /// Distribution bucket accepts new bags.
    pub accepting_new_bags: bool,

    /// Distribution bucket serves objects.
    pub distributing: bool,

    /// Pending invitations for workers to distribute the bucket.
    pub pending_invitations: BTreeSet<WorkerId>,

    /// Active operators to distribute the bucket.
    pub operators: BTreeSet<WorkerId>,

    /// Number of assigned bags.
    pub assigned_bags: u64,
}
```

The query node maintains a table of distribution operators and their invitation status. As invites are made to new operators query-nod handles `DistributionBucketOperatorInvited` and adds an operator with a reference to the bucket. On the runtime the invitation for the operator is stored in the pending_invitations set.

Currently the only way the invited operator was removed is by cancelling the invite, (where query-node handles `DistributionBucketInvitationCancelled` event).

When the lead deletes the bucket, the runtime only checks that `bucket.operators` set is empty and the bucket is deleted. When query-node handles the `DistributionBucketDeleted` event it tried to delete the bucket but there were dandling references from the invited distribution operators.

The fix basically deletes the invited operators.

**Comment by mnaamani** [comment]
The fix was done on the Query-Node
Another solution would have been to at runtime level: emit one `DistributionBucketInvitationCancelled`  for each pending invite?

**Comment by traumschule** [comment]
> I prepared a fix #3458 can you test on your node @traumschule

[works](https://discord.com/channels/811216481340751934/859752198166282280/954834805730738317)! also tried to rebase but

```
INFO: Last block: 4911693       : 30187 blocks behind
INFO: Starting the event queue                                                                                         
INFO: Processing block: 0004919489-5db65, events count: 7 
ERROR: Stopping the proccessor due to errors: {                                                                                                                                                                                                
  "message": "column Channel.active_videos_counter does not exist",
  "length": 128,                                 
  "name": "QueryFailedError",
  "severity": "ERROR",                                                                                                 
  "code": "42703",              
  "position": "783",                                                                                                                                                                                                                           
  "file": "parse_relation.c",
  "line": "3349",
  "routine": "errorMissingColumn",
  "query": "SELECT DISTINCT \"distinctAlias\".\"Channel_id\" as \"ids_Channel_id\" FROM (SELECT \"Channel\".\"id\" AS \"Channel_id\", \"Channel\".\"created_at\" AS \"Channel_created_at\", \"Channel\".\"created_by_id\" AS \"Channel_created_
by_id\", \"Channel\".\"updated_at\" AS \"Channel_updated_at\", \"Channel\".\"updated_by_id\" AS \"Channel_updated_by_id\", \"Channel\".\"deleted_at\" AS \"Channel_deleted_at\", \"Channel\".\"deleted_by_id\" AS \"Channel_deleted_by_id\", \"
Channel\".\"version\" AS \"Channel_version\", \"Channel\".\"owner_member_id\" AS \"Channel_owner_member_id\", \"Channel\".\"owner_curator_group_id\" AS \"Channel_owner_curator_group_id\", \"Channel\".\"category_id\" AS \"Channel_category_i
d\", \"Channel\".\"reward_account\" AS \"Channel_reward_account\", \"Channel\".\"title\" AS \"Channel_title\", \"Channel\".\"description\" AS \"Channel_description\", \"Channel\".\"active_videos_counter\" AS \"Channel_active_videos_counter
\", \"Channel\".\"cover_photo_id\" AS \"Channel_cover_photo_id\", \"Channel\".\"avatar_photo_id\" AS \"Channel_avatar_photo_id\", \"Channel\".\"is_public\" AS \"Channel_is_public\", \"Channel\".\"is_censored\" AS \"Channel_is_censored\", \
"Channel\".\"language_id\" AS \"Channel_language_id\", \"Channel\".\"created_in_block\" AS \"Channel_created_in_block\", \"Channel__category\".\"id\" AS \"Channel__category_id\", \"Channel__category\".\"created_at\" AS \"Channel__category_
created_at\", \"Channel__category\".\"created_by_id\" AS \"Channel__category_created_by_id\", \"Channel__category\".\"updated_at\" AS \"Channel__category_updated_at\", \"Channel__category\".\"updated_by_id\" AS \"Channel__category_updated_
by_id\", \"Channel__category\".\"deleted_at\" AS \"Channel__category_deleted_at\", \"Channel__category\".\"deleted_by_id\" AS \"Channel__category_deleted_by_id\", \"Channel__category\".\"version\" AS \"Channel__category_version\", \"Channe
l__category\".\"name\" AS \"Channel__category_name\", \"Channel__category\".\"active_videos_counter\" AS \"Channel__category_active_videos_counter\", \"Channel__category\".\"created_in_block\" AS \"Channel__category_created_in_block\" FROM
 \"channel\" \"Channel\" LEFT JOIN \"channel_category\" \"Channel__category\" ON \"Channel__category\".\"id\"=\"Channel\".\"category_id\" WHERE \"Channel\".\"id\" = $1) \"distinctAlias\" ORDER BY \"Channel_id\" ASC LIMIT 1",
  "parameters": [
    "861"
INFO: Shutting down...
  ]
} name: QueryFailedError, message: column Channel.active_videos_counter does not exist, stack: QueryFailedError: column Channel.active_videos_counter does not exist
    at new QueryFailedError (/joystream/node_modules/typeorm/error/QueryFailedError.js:12:28)
    at PostgresQueryRunner.<anonymous> (/joystream/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:248:31)
    at step (/joystream/node_modules/tslib/tslib.js:143:27) 
    at Object.throw (/joystream/node_modules/tslib/tslib.js:124:57)
    at rejected (/joystream/node_modules/tslib/tslib.js:115:69)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
ERROR: {} name: Error, message: QueryFailedError: column Channel.active_videos_counter does not exist, stack: Error: QueryFailedError: column Channel.active_videos_counter does not exist
    at MappingsProcessor.processingLoop (/joystream/node_modules/@joystream/hydra-processor/lib/process/MappingsProcessor.js:75:23)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async Promise.all (index 1)
    at async MappingsProcessor.start (/joystream/node_modules/@joystream/hydra-processor/lib/process/MappingsProcessor.js:25:9)
    at async ProcessorRunner.process (/joystream/node_modules/@joystream/hydra-processor/lib/start/ProcessorRunner.js:47:9)
    at async Run.run (/joystream/node_modules/@joystream/hydra-processor/lib/commands/run.js:27:13)
    at async Run._run (/joystream/node_modules/@oclif/command/lib/command.js:43:20)
    at async Config.runCommand (/joystream/node_modules/@oclif/config/lib/config.js:173:24)
    at async Main.run (/joystream/node_modules/@oclif/command/lib/main.js:27:9)
    at async Main._run (/joystream/node_modules/@oclif/command/lib/command.js:43:20)
INFO: Closing the database connection...
```

**Comment by mnaamani** [comment]
Yes that last error is because latest query-node on master has a new schema, so you would need to drop the processor database, and re-run the db:prepare and db:migrate scripts, then start processor which would start again at initial block for giza.
